### DISCUSSION TITLE
NNAPI usage via onnxruntime

### DISCUSSION THREAD

**Discussion by ashwinjosephk** [microsoft/onnxruntime#10692] [Other Q&A] [answered]
Hi Team,
I would like to enquire how the Onnxruntime does execution in the backend upon setting the NNAPI related flags as mentioned in the link https://onnxruntime.ai/docs/execution-providers/NNAPI-ExecutionProvider.html .
My Android device (My Android API version for the device is 30 and ABI is arm64-v8a so I guess its supported) uses a chip that can utilize CPU, GPU and DSP. My query is if I use the flag NNAPI_FLAG_CPU_DISABLED , which runtime does Onnxruntime distribute the execution to? What all are the factors involved in switching? Is it possible to set one runtime over another? Say I want to get the execution done preferably with DSP over GPU.

I also had another query if it is necessary to quantize the onnx models if I have to utilize DSP runtime via NNAPI and Onnxruntime. Other platform Runtime Platforms do instruct to do so.

**Comment by Uwenisme** [comment]
Hi, you know to select different devices now? Could you share your solutions?

**Comment by skottmckay** [comment, marked as answer]
The ORT NNAPI EP will be preferred over the ORT CPU EP when assigning nodes in the model. 

If the NNAPI EP can handle a specific operator ('handle' meaning convert to the equivalent NNAPI operator), nodes involving that operator will be assigned to the NNAPI EP. Any remaining nodes will be handled by the ORT CPU EP.

The NNAPI EP will create NNAPI model/s from the nodes assigned to it at runtime. We don't have much control over how NNAPI itself will execute the NNAPI model we create. It will internally pick whatever it thinks is best. 

Setting the NNAPI_FLAG_CPU_DISABLED flag will prevent NNAPI from running operators in the NNAPI model that only have a CPU implementation. If there are operators like that in the NNAPI model the NNAPI model creation will fail. ORT does not fall back to using the ORT CPU EP for that node, as it has already been included in the NNAPI model. 

Basically that means the NNAPI_FLAG_CPU_DISABLED flag is good for testing if the nodes assigned to NNAPI will run well (the NNAPI fallback CPU implementation is not optimized in any way). 

FWIW NNAPI performance differs dramatically across devices. Your best bet is to run the model both with the NNAPI EP enabled and disabled and pick whichever option is best on a per-device basis. 

Whether you need to quantize or not will most likely depend on the hardware in the particular device. The NNAPI spec supports both quantized and float data types in general, but the device implementation of the spec may only support quantized types. 

e.g. if the NNAPI implementation for the particular DSP only supports 8-bit data types, you'll need to quantize the model to use that. that doesn't mean NNAPI can't be used - for example the GPU on the device may handle float data types and your model could potentially be executed by NNAPI using the GPU instead of the DSP.

**Comment by grybouilli** [comment]
**UPDATE** : Also note that **NNAPI has become [deprecated](https://developer.android.com/ndk/guides/neuralnetworks/migration-guide)**. To use GPU, look into [WebGPU EP](https://github.com/microsoft/onnxruntime/issues/23565).

This is a bit old but since I landed here while looking for this, i might as well mention : while it is not possible to choose specifically which HW to run on from ONNX, **it is possible to know which HW NNAPI ends up using** .

On a adb shell : 

Before running inference, run :
```bash
setprop debug.nn.vlog "all"
```
Then run inference and :

```bash
logcat -d | grep -iE "neuralnetworks|CompilationBuilder|DeviceManager|nnapi" | grep $(pgrep your_program_running_inference)
```
(replace `your_program_running_inference` with your inference program name)

Look for lines such as :
```
ExecutionPlan: ModelBuilder::findBestDeviceForEachOperation(SLICE:0) = 0 (nnapi-reference)
04-16 16:38:35.550 24347 24347 I ExecutionPlan: ModelBuilder::findBestDeviceForEachOperation(SLICE:1) = 0 (nnapi-reference)
04-16 16:38:35.550 24347 24347 I ExecutionPlan: ModelBuilder::partitionTheWork: only one best device: 0 = nnapi-reference
04-16 16:38:35.550 24347 24347 I ExecutionPlan: ExecutionPlan::SimpleBody::finish: compilation finished successfully on nnapi-reference
```

Here I can see that SLICE:0 and SLICE:1 were run with the NNAPI back-end `nnapi-reference` which is the CPU.
NB :
```
nnapi-reference -> cpu
qti-gpu -> hardware accelator
qti-dsp -> hardware accelator
qti-hta -> hardware accelator
```
cf  : [this github issue](https://github.com/tensorflow/tensorflow/issues/43541)
